File: ref/PNU CCTV/IDIS NVR/IDIS_GDK_x86_180524/4.4.0/tester/app/tester_cpp_app_g2_search_controller/search/search_date_info.cpp
// search_date_info.cpp : implementation file
//

#include "stdafx.h"
#include "search_date_info.h"

#include <set>

using namespace client;

//////////////////////////////////////////////////////////////////////////

search_date_info::search_date_info(void)
{

}

search_date_info::~search_date_info(void)
{
    if (empty() == false) clear();
}

//////////////////////////////////////////////////////////////////////////

void search_date_info::add_rectime_info(const G2RECORD_TIME_INFO& rti)
{
    _list.push_back(rti);
}

bool search_date_info::get_date_info(int channelext, search::SEARCH_DATE_INFO_LIST& dates) const
{
    search::SEARCH_DATE_INFO_LIST().swap(dates);
    if (empty()) return false;

    int i = 0;
    size_t j = 0;

    std::set<CTime> buffer;
    CTime time;
    const CTimeSpan span(1, 0, 0, 0);

    // request about tango channel (channelext)
    if (client::valid_channel_ext(channelext)) {
        for (search::RECORD_TIME_INFO_LIST::const_iterator itr(_list.begin());
             itr != _list.end();
             ++itr) {
            time = CTime(g2_time_to_time32_t(&itr->_spot._time));

            for (i = 1; i < itr->_time_size; ++i) {
                for (j = 0; j < itr->_channels._len; ++j) {
                    const G2RECORD_TYPE_INFO::ELEMENT_TYPE& element = itr->_record_type[i]._elements[j];
                    if (element._channelext == channelext &&
                        element._rec_type != 0) {
                        buffer.insert(time);
                    }
                }
                time += span;
            }
        }
    }
    else {
        for (search::RECORD_TIME_INFO_LIST::const_iterator itr(_list.begin());
             itr != _list.end();
             ++itr) {
            time = CTime(g2_time_to_time32_t(&itr->_spot._time));

            for (i = 1; i < itr->_time_size; ++i) {
                for (j = 0; j < itr->_channels._len; ++j) {
                    const G2RECORD_TYPE_INFO::ELEMENT_TYPE& element = itr->_record_type[i]._elements[j];
                    if (element._rec_type != 0) {
                        buffer.insert(time);
                    }
                }
                time += span;
            }
        }
    }

    dates.insert(dates.end(), buffer.begin(), buffer.end());

    return (dates.empty() != true);
}

File: ref/PNU CCTV/IDIS NVR/G2ClientGDK_x64_4.3.2_rev_01_170630/4.3.2/tester/csharp/tester_screen/search/search_date_info.cs
using System;
using System.Collections.Generic;
using System.Text;
using GDK;

namespace GDK_tester
{
    using G2RECORD_TIME_INFO_LIST = LinkedList<G2RECORD_TIME_INFO>;

    public class search_date_info
    {
        public search_date_info()
        {
            _list = new G2RECORD_TIME_INFO_LIST();
        }

        public void clear()
        {
            _list = new G2RECORD_TIME_INFO_LIST();
        }
        public bool empty()
        {
            return (_list.Count == 0);
        }
        public void append(ref G2RECORD_TIME_INFO rti)
        {
            lock (_list)
            {
                _list.AddLast(rti);
            }
        }

        public bool get(int channelext, List<DateTime> dates)
        {
            if (empty()) return false;

            TimeSpan span = new TimeSpan(1, 0, 0, 0);

            lock (_list)
            {
                foreach (G2RECORD_TIME_INFO rti in _list)
                {
                    if (rti._spot._time.valid)
                    {
                        DateTime time = rti._spot._time;
                        for (int i = 1; i < rti._time_size; ++i)
                        {
                            for (int j = 0; j < rti._channels._len; ++j)
                            {
                                G2RECORD_TYPE_INFO.ELEMENT_TYPE e = rti._record_type[i]._elements[j];
                                if (e._rec_type != 0)
                                {
                                    if (channelext == -1 ||
                                        channelext == e._channelext)
                                    {
                                        dates.Add(time);
                                    }
                                }
                            }
                            time += span;
                        }
                    }
                }

                dates.Sort();

                for (int i = 0; i < dates.Count - 1; )
                {
                    if (dates[i] == dates[i + 1])
                    {
                        dates.RemoveAt(i);
                    }
                    else
                    {
                        ++i;
                    }
                }
                return (dates.Count != 0);
            }
        }
        public bool get(List<DateTime> dates)
        {
            return get(-1, dates);
        }

        public G2RECORD_TIME_INFO_LIST _list;
    }
}
